### DISCUSSION TITLE
Did something change recently to make lock files more strict? Docker build failing on macOS host

### DISCUSSION THREAD

**Discussion by anentropic** [pdm-project/pdm#3173] [Q&A] [answered]
Until today I'd been able to run `pdm lock` on my local macOS and then `COPY pdm.lock` in a Dockerfile and install packages as shown here https://pdm-project.org/latest/usage/advanced/#use-pdm-in-a-multi-stage-dockerfile

But today I started getting this error:
```
 > [fargate-task 6/6] RUN pdm install --prod --skip=:all --fail-fast --check --prod --no-editable:
0.753 WARNING: Project requires a python version of ~=3.11, The virtualenv is being created for you as it cannot be matched to the right version.
0.753 INFO: python.use_venv is on, creating a virtualenv for this project...
0.866 Virtualenv is created successfully at /usr/src/app/.venv
0.892 STATUS: Resolving packages from lockfile...
0.916 ERROR: None of the lock targets matches the current env (==3.11.5, manylinux_2_36_aarch64, cpython):
0.916  - (==3.11.5, macos_14_0_arm64, cpython)
0.916 [PdmException]: No compatible lock target found
0.916 WARNING: Add '-v' to see the detailed traceback
```

It appears from the error that problem is a mismatch in the platform `manylinux_2_36_aarch64` vs `macos_14_0_arm64`

Does this mean it's no longer possible to `pdm lock` in my macos terminal and then use the lock file in a linux docker build?

Reading https://pdm-project.org/en/latest/usage/lock-targets/ it sounds like something has gone wrong with `pdm lock`

> By default, PDM will try to make a lock file that works on all platforms

But the lock file I got from running just `pdm lock` contains:
```
[[metadata.targets]]
requires_python = "==3.11.5"
platform = "macos_14_0_arm64"
implementation = "cpython"
gil_disabled = false
```

If I do an explicit `pdm lock --platform manylinux_2_36_aarch64` from macOS side then the Docker build succeeds

**Comment by anentropic** [comment]
if I subsequently `pdm lock` without specifying then it locks for `manylinux_2_36_aarch64` again

if I delete the lock file and `pdm lock` then I get back to an "any platform" lock file having just:
```
[[metadata.targets]]
requires_python = "~=3.11"
```

which is good, and the Docker build succeeds again

I'm a bit confused how I ended up with the `macos_14_0_arm64` platform specifier in the first place as I definitely didn't explicitly request it

**Comment by frostming** [comment, marked as answer]
The `metadata.targets` introduced in lock version 4.5.0 are translated depending on whether `cross_platform` strategy is present in lock version < 4.5.0.

If the lock file doesn't contain `cross_platform` strategy, the lock file is a **strict** lock file that works on the current environment only. We just keep the same after migrating to 4.5.0, by specifying the platform and implementation in metadata.targets.
